pkg/kubelet/cm/container_manager_linux.go

```diff
@@ -1060,7 +1060,32 @@ func isKernelPid(pid int) bool {
 	return err != nil && os.IsNotExist(err)
 }
 
+// GetCapacity returns node capacity data for "cpu", "memory", "ephemeral-storage", and "huge-pages*"
+// At present this method is only invoked when introspecting ephemeral storage
```

Comment:
Added this comment to emphasize the practical need to work harder to deliver ephemeral-storage data: This `GetCapacity` method is *only* called to get ephemeral storage data. See here where we refer to it when defining the set of `MachineInfo` setter functions:

https://github.com/kubernetes/kubernetes/blob/v1.21.0/pkg/kubelet/kubelet_node_status.go#L617

Those set of funcs are then assigned to the main kubelet instance:

https://github.com/kubernetes/kubernetes/blob/v1.21.0/pkg/kubelet/kubelet.go#L848

And then those status funcs are invoked every time kubelet sets the node status for a node:

https://github.com/kubernetes/kubernetes/blob/v1.21.0/pkg/kubelet/kubelet_node_status.go#L583

What's interesting here is that when we invoke `MachineInfo` with `GetCapacity()` as the `capacityFunc`, we see that we only use `capacityFunc` to derive ephemeral storage data, and nothing else:

https://github.com/kubernetes/kubernetes/blob/v1.21.0/pkg/kubelet/nodestatus/setters.go#L332

Thus, we should improve the resiliency/reliability of `GetCapacity()` to actually return real ephemeral storage allocatable data.

Reply:
so where cpu and memory is retrieved from?

Reply:
CPU and memory are stored in the containerManager instance's `capacity` property _when containerManager is instantiated_, right here:

https://github.com/kubernetes/kubernetes/blob/v1.21.0/pkg/kubelet/cm/container_manager_linux.go#L246

However, storage is _not_ stored at that time, which is why the race condition occurs, and why we may need to call it "just in time" during node info retrieval/node registration. (See this comment: https://github.com/kubernetes/kubernetes/blob/v1.21.0/pkg/kubelet/cm/container_manager_linux.go#L245)

CPU/memory info is gathered using another function, see here:

https://github.com/kubernetes/kubernetes/blob/v1.21.0/pkg/kubelet/nodestatus/setters.go#L296

there are 2 tl;dr points:

1. storage and CPU/memory are retrieved discretely, using individual code paths
2. CPU/memory is gathered immediately when the kubelet runtime starts, and has its own error handling (see code); storage is gathered a little later than CPU/memory, and thus node info events *may* occur before storage has been gathered, leading to node registrations with "0" allocatable ephemeral-storage.

This PR accomplishes the following:

If we detect that allocatable ephemeral-storage has not yet been gathered during a node info/node registration operation, we attempt to eagerly gather that storage data "just in time" in order to prevent registering a node with "0" allocatable ephemeral-storage.

Hope that makes sense!